Hello! We are going to use a chunk of Sloan Digital Sky Survey's data here. Our aim is to build some machine learning model to classify galaxies/stars/quasars based on their spectral properties. After that we will also consider differnt metrics to assess those models' performance.

First thing first. Let's import some mandatory modules and packages.

In [1]:
import warnings
warnings.filterwarnings('ignore')
from IPython.display import display
from matplotlib import pyplot as plt
from matplotlib import style
import numpy as np
import os
import pandas as pd
import seaborn as sns
style.use(style='seaborn-deep')
from tabulate import tabulate
import pickle
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier, GradientBoostingClassifier, ExtraTreesClassifier, BaggingClassifier
from sklearn.neural_network import MLPClassifier
from xgboost import XGBClassifier

Accessing the dataset..............

In [2]:
sdss_df = pd.read_csv('star_classification.csv')

Let's check some basic info about the dataset.

In [3]:
sdss_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 18 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   obj_ID       100000 non-null  float64
 1   alpha        100000 non-null  float64
 2   delta        100000 non-null  float64
 3   u            100000 non-null  float64
 4   g            100000 non-null  float64
 5   r            100000 non-null  float64
 6   i            100000 non-null  float64
 7   z            100000 non-null  float64
 8   run_ID       100000 non-null  int64  
 9   rerun_ID     100000 non-null  int64  
 10  cam_col      100000 non-null  int64  
 11  field_ID     100000 non-null  int64  
 12  spec_obj_ID  100000 non-null  float64
 13  class        100000 non-null  object 
 14  redshift     100000 non-null  float64
 15  plate        100000 non-null  int64  
 16  MJD          100000 non-null  int64  
 17  fiber_ID     100000 non-null  int64  
dtypes: float64(10), int64(7),

So, there are 100,000 entries or rows and 18 columns in total.

In [4]:
print("The shape of the dataset: {}".format(sdss_df.shape))

The shape of the dataset: (100000, 18)


In [5]:
classes = sdss_df['class'].unique()
print(classes)

['GALAXY' 'QSO' 'STAR']


In [6]:
sdss_df

obj_ID       alpha      delta         u         g         r   
0      1.240000e+18  135.689107  32.494632  23.87882  22.27530  20.39501  \
1      1.240000e+18  144.826101  31.274185  24.77759  22.83188  22.58444   
2      1.240000e+18  142.188790  35.582444  25.26307  22.66389  20.60976   
3      1.240000e+18  338.741038  -0.402828  22.13682  23.77656  21.61162   
4      1.240000e+18  345.282593  21.183866  19.43718  17.58028  16.49747   
...             ...         ...        ...       ...       ...       ...   
99995  1.240000e+18   39.620709  -2.594074  22.16759  22.97586  21.90404   
99996  1.240000e+18   29.493819  19.798874  22.69118  22.38628  20.45003   
99997  1.240000e+18  224.587407  15.700707  21.16916  19.26997  18.20428   
99998  1.240000e+18  212.268621  46.660365  25.35039  21.63757  19.91386   
99999  1.240000e+18  196.896053  49.464643  22.62171  21.79745  20.60115   

              i         z  run_ID  rerun_ID  cam_col  field_ID   spec_obj_ID   
0      19.16573  18.79371    3606       301        2        79  6.540000e+18  \
1      21.16812  21.61427    4518       301        5       119  1.180000e+19   
2      19.34857  18.94827    3606       301        2       120  5.150000e+18   
3      20.50454  19.25010    4192       301        3       214  1.030000e+19   
4      15.97711  15.54461    8102       301        3       137  6.890000e+18   
...         ...       ...     ...       ...      ...       ...           ...   
99995  21.30548  20.73569    7778       301        2       581  1.060000e+19   
99996  19.75759  19.41526    7917       301        1       289  8.590000e+18   
99997  17.69034  17.35221    5314       301        4       308  3.110000e+18   
99998  19.07254  18.62482    3650       301        4       131  7.600000e+18   
99999  20.00959  19.28075    3650       301        4        60  8.340000e+18   

        class  redshift  plate    MJD  fiber_ID  
0      GALAXY  0.634794   5812  56354       171  
1      GALAXY  0.779136  10445  58158       427  
2      GALAXY  0.644195   4576  55592       299  
3      GALAXY  0.932346   9149  58039       775  
4      GALAXY  0.116123   6121  56187       842  
...       ...       ...    ...    ...       ...  
99995  GALAXY  0.000000   9374  57749       438  
99996  GALAXY  0.404895   7626  56934       866  
99997  GALAXY  0.143366   2764  54535        74  
99998  GALAXY  0.455040   6751  56368       470  
99999  GALAXY  0.542944   7410  57104       851  

[100000 rows x 18 columns]

Let's check if there is any null value.

In [7]:
sdss_df.isnull().sum()

obj_ID         0
alpha          0
delta          0
u              0
g              0
r              0
i              0
z              0
run_ID         0
rerun_ID       0
cam_col        0
field_ID       0
spec_obj_ID    0
class          0
redshift       0
plate          0
MJD            0
fiber_ID       0
dtype: int64


All okay! No null value. Amazing!


In [8]:
sdss_df['class'].value_counts()

class
GALAXY    59445
STAR      21594
QSO       18961
Name: count, dtype: int64

So, in our dataset there are 59445 galaxies, 21594 star, and 18961 QSOs.